packages/apollo-engine-reporting/src/schemaReporter.ts

```diff
@@ -0,0 +1,190 @@
+    this.executableSchemaDocument = schemaSdl;
+    this.isStopped = false;
+  }
+
+  public stopped(): Boolean {
+    return this.isStopped;
+  }
+
+  public stop() {
+    this.isStopped = true;
+  }
+
+  public async reportServerInfo(
+    withExecutableSchema: boolean,
+  ): Promise<ReportServerInfoReturnVal> {
+    const { data, errors } = await this.graphManagerQuery({
+      info: this.serverInfo,
+      executableSchema: withExecutableSchema
+        ? this.executableSchemaDocument
+        : null,
+    });
+
+    if (errors) {
+      throw new Error((errors || []).map((x: any) => x.message).join('\n'));
+    }
+
+    function msgForUnexpectedResponse(data: any): string {
+      return [
+        'Unexpected response shape from Apollo Graph Manager when',
+        'reporting server information for schema reporting. If',
+        'this continues, please reach out to support@apollographql.com.',
+        'Received response:',
+        JSON.stringify(data),
+      ].join(' ');
+    }
+
+    if (!data || !data.me || !data.me.__typename) {
+      throw new Error(msgForUnexpectedResponse(data));
+    }
+
+    if (data.me.__typename === 'UserMutation') {
+      this.isStopped = true;
+      throw new Error(
+        [
+          'This server was configured with an API key for a user.',
+          "Only a service's API key may be used for schema reporting.",
+          'Please visit the settings for this graph at',
+          'https://engine.apollographql.com/ to obtain an API key for a service.',
+        ].join(' '),
+      );
+    } else if (data.me.__typename === 'ServiceMutation') {
+      if (!data.me.reportServerInfo) {
+        throw new Error(msgForUnexpectedResponse(data));
+      }
+      return data.me.reportServerInfo;
+    } else {
+      throw new Error(msgForUnexpectedResponse(data));
+    }
+  }
+
+  private async graphManagerQuery(
+    variables: ReportServerInfoVariables,
+  ): Promise<SchemaReportingServerInfoResult> {
+    const request: GraphQLRequest = {
+      query: reportServerInfoGql,
+      operationName: 'ReportServerInfo',
+      variables: variables,
+    };
+    const httpRequest = new Request(this.url, {
+      method: 'POST',
+      headers: this.headers,
+      body: JSON.stringify(request),
+    });
+
+    const httpResponse = await fetch(httpRequest);
+
+    if (!httpResponse.ok) {
+      throw new Error([
+        `An unexpected HTTP status code (${httpResponse.status}) was`,
+        'encountered during schema reporting.'
+      ].join(' '));
+    }
+
+    try {
+      // JSON parsing failure due to malformed data is the likely failure case
+      // here.  Any non-JSON response (e.g. HTML) is usually the suspect.
+      return await httpResponse.json();
+    } catch (error) {
+      throw new Error(
+        [
+          "Couldn't report server info to Apollo Graph Manager.",
+          'Parsing response as JSON failed.',
+          'If this continues please reach out to support@apollographql.com',
```

Comment:
Putting just a space between our email address and the error will read strangely. Newline? Period?